src/sentry/analytics/event.py

```diff
@@ -70,11 +70,7 @@ def extract(self, value):
 class Event(object):
     __slots__ = ["uuid", "attributes", "data", "datetime", "type"]
 
-    type = None
-
-    attributes = ()
-
-    def __init__(self, type=None, datetime=None, **items):
+    def __init__(self, type=None, datetime=None, attributes=(), **items):
```

Comment:
https://docs.python.org/3/reference/datamodel.html#notes-on-using-slots

> __slots__ are implemented at the class level by creating descriptors (Implementing Descriptors) for each variable name. As a result, class attributes cannot be used to set default values for instance variables defined by __slots__; otherwise, the class attribute would overwrite the descriptor assignment.

So to fix that, `attributes = ()` needs to be moved into the constructor as a default value. `type = None` is redundant.

Reply:
Actually now that I'm looking more into this, we probably need attributes to be a class variable. I see we have `cls.attribute` in a classmethod, and a bunch of event subclasses override this class variable.

Reply:
Slots is a good memory optimization here, so I doubt we'd want to remove it, but perhaps we could remove attributes from slots? I don't know enough about classes in python yet, but that seems like a safe change...?

Reply:
i dont think attributes being in slots is correct, if that helps

Reply:
(if it is indeed a class variable)

Reply:
It is, and yeah that helps - I think removing attributes from slots is the right move here. I don't think what I did is really correct. (Hard to tell if it regresses any tests on 3.6 because of all the failure kludge right now)